homeassistant/components/upnp/sensor.py

```diff
@@ -1,275 +1,257 @@
+    ) -> None:
         """Initialize the base sensor."""
+        self._coordinator = coordinator
         self._device = device
+        self._sensor_type = sensor_type
 
-    async def async_added_to_hass(self):
-        """Subscribe to sensors events."""
-        self.async_on_remove(
-            async_dispatcher_connect(
-                self.hass, SIGNAL_REMOVE_SENSOR, self._upnp_remove_sensor
-            )
+    @property
+    def should_poll(self) -> bool:
+        """Inform we should not be polled."""
+        return False
+
+    @property
+    def icon(self) -> str:
+        """Icon to use in the frontend, if any."""
+        return "mdi:server-network"
+
+    @property
+    def available(self) -> bool:
+        """Return if entity is available."""
+        device_value_key = self._sensor_type["device_value_key"]
+        return (
+            self._coordinator.last_update_success
+            and device_value_key in self._coordinator.data
+            and not isinstance(self._coordinator.data[device_value_key], Exception)
         )
 
-    @callback
-    def _upnp_remove_sensor(self, device):
-        """Remove sensor."""
-        if self._device != device:
-            # not for us
-            return
+    @property
+    def name(self) -> str:
+        """Return the name of the sensor."""
+        return f"{self._device.name} {self._sensor_type['name']}"
 
-        self.hass.async_create_task(self.async_remove())
+    @property
+    def unique_id(self) -> str:
+        """Return an unique ID."""
+        return f"{self._device.udn}_{self._sensor_type['unique_id']}"
 
     @property
-    def device_info(self):
+    def unit_of_measurement(self) -> str:
+        """Return the unit of measurement of this entity, if any."""
+        return self._sensor_type["unit"]
+
+    @property
+    def device_info(self) -> Mapping[str, any]:
         """Get device info."""
         return {
             "connections": {(dr.CONNECTION_UPNP, self._device.udn)},
-            "identifiers": {(DOMAIN_UPNP, self._device.udn)},
             "name": self._device.name,
             "manufacturer": self._device.manufacturer,
             "model": self._device.model_name,
         }
 
+    async def async_added_to_hass(self) -> None:
+        """Subscribe to sensors events."""
+        self._coordinator.async_add_listener(self.async_write_ha_state)
 
-class RawUPnPIGDSensor(UpnpSensor):
-    """Representation of a UPnP/IGD sensor."""
-
-    def __init__(self, device, sensor_type_name, sensor_type):
-        """Initialize the UPnP/IGD sensor."""
-        super().__init__(device)
-        self._type_name = sensor_type_name
-        self._type = sensor_type
-        self._name = "{} {}".format(device.name, sensor_type["name"])
-        self._state = None
+        self.async_on_remove(
+            async_dispatcher_connect(
+                self.hass, SIGNAL_REMOVE_DEVICE, self._async_upnp_remove_sensor
+            )
+        )
 
```

Comment:
If we add an update method here that awaits coordinator.async_request_refresh, we'll allow users to schedule manual updates using the common entity update service.
```py
async def async_update(self):
    """Request an update."""
    await self._coordinator.async_request_refresh()
```